Ravendb 6013

aviv86 opened the pull request.

ppekrol commented on src/Raven.Client/Documents/Session/StreamQueryStatistics.cs:
```diff
@@ -0,0 +1,37 @@
+        public int TotalResults { get; set; }
+        public long ResultEtag { get; set; }
+
+        internal void UpdateQueryStats(string indexName, string isStale, string indexTimestamp, string totalResults, string resultEtag)
```
I think you can safely move everything from here to EnsureValidStreamQueryStats, this is also a part of validation checking if the type of value is parsable to int for example

ppekrol commented on src/Raven.Server/Documents/Queries/StreamDocumentQueryResult.cs:
```diff
@@ -75,18 +75,35 @@ public void Dispose()
+            _writer.WriteComma();
+
+            _writer.WritePropertyName(nameof(StreamQueryStatistics.IsStale));
+            _writer.WriteString(IsStale ? "true" : "false");
```
WriteBool ?

ppekrol commented on src/Raven.Server/Documents/Queries/StreamDocumentQueryResult.cs:
```diff
@@ -75,18 +75,35 @@ public void Dispose()
+            _writer.WriteComma();
+
+            _writer.WritePropertyName(nameof(StreamQueryStatistics.TotalResults));
+            _writer.WriteString(TotalResults.ToInvariantString());
```
WriteInteger

ppekrol commented on src/Raven.Server/Documents/Queries/StreamDocumentQueryResult.cs:
```diff
@@ -75,18 +75,35 @@ public void Dispose()
-            _response.Headers.Add("Raven-Total-Results", TotalResults.ToInvariantString());
-            _response.Headers.Add("Raven-Index-Timestamp", IndexTimestamp.GetDefaultRavenFormat(isUtc: true));
+            _writer.WritePropertyName(nameof(StreamQueryStatistics.ResultEtag));
+            _writer.WriteString(ResultEtag.ToString());
```
WriteInteger

ppekrol commented on src/Raven.Client/Documents/Session/Operations/StreamOperation.cs:
```diff
@@ -112,9 +114,13 @@ public IEnumerator<BlittableJsonReaderObject> SetResult(StreamResult response)
-                if (string.Equals(property, "Results") == false)
-                    UnmanagedJsonParserHelper.ThrowInvalidJson();
 
+                if (string.Equals(property, nameof(StreamQueryStatistics.ResultEtag)))
```
Do not check if property is ResultEtag, do it inside the EnsureValidStreamQueryStats

ppekrol commented on src/Raven.Client/Documents/Session/Operations/StreamOperation.cs:
```diff
@@ -126,6 +132,37 @@ public IEnumerator<BlittableJsonReaderObject> SetResult(StreamResult response)
+            var indexTimestamp = UnmanagedJsonParserHelper.ReadString(context, response.Stream, parser, state, buffer);
+
+            property = UnmanagedJsonParserHelper.ReadString(context, response.Stream, parser, state, buffer);
+            if (string.Equals(property, "Results") == false)
```
Move this outside this method, this is not part of the stats

ppekrol commented on src/Raven.Client/Documents/Session/Operations/StreamOperation.cs:
```diff
@@ -198,7 +235,11 @@ public void Dispose()
                 var property = UnmanagedJsonParserHelper.ReadString(_session.Context, _response.Stream, _parser, _state, _buffer);
-                if (string.Equals(property, "Results") == false)
+
+                if (string.Equals(property, nameof(StreamQueryStatistics.ResultEtag)))
```
Same here

ppekrol commented on src/Raven.Client/Documents/Session/DocumentSession.Stream.cs:
```diff
@@ -21,33 +21,57 @@ public IEnumerator<StreamResult<T>> Stream<T>(IQueryable<T> query)
+                yield return CreateStreamResult<T>(json, projections);
             }
+
+            enumerator.Dispose();
```
wrap in using

ppekrol commented on src/Raven.Client/Documents/Session/Operations/StreamOperation.cs:
```diff
@@ -13,6 +13,7 @@ namespace Raven.Client.Documents.Session.Operations
     internal class StreamOperation
     {
         private readonly InMemoryDocumentSessionOperations _session;
+        public StreamQueryStatistics Statistics;
```
can you mark it as readonly and pass it using the constructor?

aviv86 committed:
RavenDB-6013 :
- remove headers from query streaming, write those values as additional json properties
- add Stream() overload with an out parameter to retrieve the StreamQueryStatistics
